okhttp/src/main/java/okhttp3/internal/connection/RealConnection.kt

```diff
@@ -575,10 +576,17 @@ class RealConnection(
       return true // Host match. The URL is supported.
     }
 
-    // We have a host mismatch. But if the certificate matches, we're still good.
-    return !noCoalescedConnections &&
-        handshake != null &&
-        OkHostnameVerifier.verify(url.host, handshake!!.peerCertificates[0] as X509Certificate)
+    try {
+      // We have a host mismatch. But if the certificate matches, we're still good.
+      return !noCoalescedConnections &&
+          handshake != null &&
+          OkHostnameVerifier.verify(url.host, handshake!!.peerCertificates[0] as X509Certificate)
+    } catch (_: SSLPeerUnverifiedException) {
+      // OkHostnameVerifier isn't guaranteed to work if user has disabled security via
+      // TrustManager and hostnameVerifier
+      noCoalescedConnections = true
```

Comment:
@swankjesse This is only used when we already have the connectionPool lock.  calling the noCoalescedConnections method would fail because of this.  But checking here and failing if we don't have it seems problematic in an exception handler.

Thoughts?

Reply:
Triggers

```
    java.lang.AssertionError: Thread Test worker MUST hold lock on okhttp3.internal.connection.RealConnectionPool@56c6ffac
        at okhttp3.internal.connection.RealConnection.supportsUrl(RealConnection.kt:770)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:87)
        at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:34)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:100)
        at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:82)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:100)
        at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:83)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:100)
        at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:74)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:100)
        at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:197)
        at okhttp3.internal.connection.RealCall.execute(RealCall.kt:148)
        at okhttp3.CacheTest.get(CacheTest.java:2490)
        at okhttp3.CacheTest.varyMatchesAddedRequestHeaderField(CacheTest.java:1605)
```